natlib/svcd.c

```diff
@@ -183,5 +190,38 @@ static int svcd_init( lua_State *L )
         lua_call(L, 3, 0);
     }
 
+    
+    return 0;
+}
+
+// Lua: storm.n.ndispatch ( pay, srcip, srcport )
+// Dispatches an incoming notification
+
+static int svcd_ndispatch( lua_State *L )
+{
+    // Checks to make sure it is passed 3 parameters
+    if (lua_gettop(L) != 3) return luaL_error(L, "Expected (pay, srcip, srcport)"); 
+
+    size_t parlen;
+    const char* pay = lua_tolstring(L, -1, &parlen); 
+    uint16_t ivkid = lua_tonumber(L, -1); // changes pay parameter to number and returns
+    lua_getglobal(L, "SVCD"); // gets table
+    lua_pushstring(L, "oursubs"); // use oursubs as the key
+    lua_gettable(L, 5);
+    char ivkidstr[16];
+    sprintf(ivkidstr, "%d", ivkid);
+    lua_pushstring(L, ivkidstr);
+    lua_gettable(L, 6);
+
+    size_t size;
+    const char* item = lua_tolstring(L, 6, &size);
+    if (!lua_isnil(L, 7)) {
+	char newstr[3];
+	strncpy(newstr, pay, 3);
```

Comment:
char * strncpy ( char * destination, const char * source, size_t num ); copies the first num characters from source to destination. 

string.sub (s, i [, j]) returns the substring of s from the i-th character to the j-th character. If j is missing, it will return from the i-th character to the end of the string. 

After line 220, your [newstr] is the first 3 characters of the string [pay]. However, in the original code, it requires all characters of [pay] from 3rd and forward. This may cause some problem. 

Untitled